exercises/book-store/example.rs

```diff
@@ -0,0 +1,189 @@
+        //    then move the last item from the most populous group into a new group, alone,
+        //    and return
+        println!("DecomposeGroups.next() --> {:?}", self.next);
+        let return_value = self.next.clone();
+        if let Some(groups) = mem::replace(&mut self.next, None) {
+            if !(groups.is_empty() || groups.iter().all(|g| g.0.borrow().len() == 1)) {
+                let mut hypothetical;
+                for mpg_book in groups[0].0.borrow().iter() {
+                    for (idx, other_group) in groups[1..].iter().enumerate() {
+                        if !other_group.0.borrow().contains(mpg_book) {
+                            hypothetical = groups.clone();
+                            hypothetical[0].0.borrow_mut().remove(mpg_book);
+                            hypothetical[1 + idx].0.borrow_mut().insert(*mpg_book);
+                            hypothetical.sort();
+                            let hypothetical_hash = hash_of(&hypothetical);
+                            if !self.prev_states.contains(&hypothetical_hash) {
+                                self.prev_states.insert(hypothetical_hash);
+                                mem::replace(&mut self.next, Some(hypothetical));
+                                return return_value;
+                            }
+                        }
+                    }
+                }
+                // we've gone through all the items of the most populous group,
+                // and none of them can be added to any other existing group.
+                // We need to create a new group;
+                let book = {
+                    let backing_bt = groups[0].0.borrow();
+                    let mut book_iter = backing_bt.iter();
+                    book_iter.next().unwrap().clone()
+                };
+                hypothetical = groups.clone();
+                hypothetical[0].0.borrow_mut().remove(&book);
+                hypothetical.push(Group::new_containing(book));
+                hypothetical.sort();
+                self.prev_states.insert(hash_of(&hypothetical));
+                mem::replace(&mut self.next, Some(hypothetical));
+            }
+        }
+        return_value
+    }
+}
+
+impl DecomposeGroups {
+    fn new(books: &[Book]) -> DecomposeGroups {
+        let mut book_groups = GroupedBasket::new();
+        'nextbook: for book in books {
+            for idx in 0..book_groups.len() {
+                if !book_groups[idx].0.borrow().contains(&book) {
+                    book_groups[idx].0.borrow_mut().insert(*book);
+                    continue 'nextbook;
+                }
+            }
+            // if we're here, we still haven't found a place for the book.
+            // better add it to a new group
+            book_groups.push(Group::new_containing(*book));
+        }
+        book_groups.sort();
+        println!("DecomposeGroups::new() --> {:?}", book_groups);
```

Comment:
still need this?

Reply:
No, but it does get captured by the test runner, so I didn't bother to take it out. That said, I don't object to removing it.